demo-app/src/main.rs

```diff
@@ -0,0 +1,253 @@
+use std::marker::PhantomData;
+
+use borsh::{BorshDeserialize, BorshSerialize};
+use example_election::Election;
+// Items that should go in prelude
+use jmt::SimpleHasher;
+use sov_modules_api::{
+    mocks::{MockContext, MockPublicKey},
+    Context, DispatchCall, DispatchQuery, Genesis, Module, Spec,
+};
+use sov_modules_macros::{DispatchCall, DispatchQuery, Genesis, MessageCodec};
+use sov_state::{JmtStorage, Storage};
+use sovereign_sdk::{
+    core::{
+        mocks::MockProof,
+        traits::{BatchTrait, CanonicalHash, TransactionTrait},
+    },
+    jmt,
+    serial::Decode,
+    stf::{ConsensusSetUpdate, OpaqueAddress, StateTransitionFunction},
+};
+
+// we should re-export anyhow from the modules_api
+
+#[derive(Debug, PartialEq, Eq, Clone, BorshDeserialize, BorshSerialize)]
+pub struct DemoTransaction<C: Context> {
+    mock_signature: MockSignature,
+    msg: Vec<u8>,
+    phantom_context: PhantomData<C>,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, BorshDeserialize, BorshSerialize)]
+pub struct MockSignature {
+    pub_key: MockPublicKey,
+    msg_hash: [u8; 32],
+}
+
+pub struct Demo {
+    current_storage: JmtStorage,
+}
+
+impl StateTransitionFunction for Demo {
+    type StateRoot = jmt::RootHash;
+
+    type ChainParams = ();
+
+    type Transaction = DemoTransaction<MockContext>;
+
+    type Batch = Batch<MockContext>;
+
+    type Proof = MockProof;
+
+    type MisbehaviorProof = ();
+
+    fn init_chain(&mut self, _params: Self::ChainParams) {
+        Runtime::<MockContext>::genesis(self.current_storage.clone())
+            .expect("module initialization must succeed");
+        self.current_storage.finalize();
+    }
+
+    fn begin_slot(&self) {}
+
+    fn apply_batch(
+        &self,
+        batch: Self::Batch,
+        sequencer: &[u8],
+        _misbehavior_hint: Option<Self::MisbehaviorProof>,
+    ) -> Result<
+        Vec<Vec<sovereign_sdk::stf::Event>>,
+        sovereign_sdk::stf::ConsensusSetUpdate<OpaqueAddress>,
+    > {
+        let mut storage = self.current_storage.clone();
+        let mut events = Vec::new();
+        for tx in batch.take_transactions() {
+            // Do mock signature verification
+            // We just check that the signature hash matches the tx hash
+            let expected_hash = tx.mock_signature.msg_hash;
+            let found_hash = <MockContext as Spec>::Hasher::hash(&tx.msg);
```

Comment:
non-blocking:
This logic is common for the native and zk logic so we should pass Context from the outside instead of hardcoding MockContext (we can parametric Demo struct by C)

Reply:
Hmm, yeah this is a good point. Can we get away with using a concrete Spec instead of generic context here?

Reply:
If the `ZkHasher` can have a different implementation than `NativeHasher` then it looks like we have to be generic...

Reply:
Yep, that makes sense. I think I'd lean towards abstracting the difference between zk and native hashing inside the `Hasher` using conditional compilation. What do you think?

Reply:
We can do this or  do the conditional compilation one level higher (here):
Context = MockContext or Context=ZkContext and hardcode Context everywhere. Let's merge it as it is and decide later.